pkg/scheduler/internal/cache/cache.go

```diff
@@ -230,28 +253,66 @@ func (cache *schedulerCache) UpdateNodeInfoSnapshot(nodeSnapshot *nodeinfosnapsh
 	}
 
 	if len(nodeSnapshot.NodeInfoMap) > len(cache.nodes) {
-		for name := range nodeSnapshot.NodeInfoMap {
-			if _, ok := cache.nodes[name]; !ok {
-				delete(nodeSnapshot.NodeInfoMap, name)
-			}
-		}
+		cache.removeDeletedNodesFromSnapshot(nodeSnapshot)
+		updateAllLists = true
 	}
 
-	// Take a snapshot of the nodes order in the tree
+	if updateAllLists || updateNodesHavePodsWithAffinity {
+		cache.updateNodeInfoSnapshotList(nodeSnapshot, updateAllLists)
+	}
+
+	if len(nodeSnapshot.NodeInfoList) != len(nodeSnapshot.NodeInfoMap) {
+		err := fmt.Errorf("snapshot state is not consistent, length of NodeInfoList=%v not equal to length of NodeInfoMap=%v "+
+			"length of nodes in cache=%v, length of nodes in tree=%v"+
+			", trying to recover",
+			len(nodeSnapshot.NodeInfoList), len(nodeSnapshot.NodeInfoMap),
+			len(cache.nodes), cache.nodeTree.numNodes)
+		// We will try to recover by re-creating the lists for the next scheduling cycle, but still return an
+		// error to surface the problem, the error will likely cause a failure to the current scheduling cycle.
+		cache.updateNodeInfoSnapshotList(nodeSnapshot, true)
+		return err
+	}
+
+	return nil
+}
+
+func (cache *schedulerCache) updateNodeInfoSnapshotList(nodeSnapshot *nodeinfosnapshot.Snapshot, updateAll bool) {
 	nodeSnapshot.NodeInfoList = make([]*schedulernodeinfo.NodeInfo, 0, cache.nodeTree.numNodes)
-	nodeSnapshot.HavePodsWithAffinityNodeInfoList = make([]*schedulernodeinfo.NodeInfo, 0, cache.nodeTree.numNodes)
-	for i := 0; i < cache.nodeTree.numNodes; i++ {
-		nodeName := cache.nodeTree.next()
-		if n := nodeSnapshot.NodeInfoMap[nodeName]; n != nil {
-			nodeSnapshot.NodeInfoList = append(nodeSnapshot.NodeInfoList, n)
+	if updateAll {
+		// Take a snapshot of the nodes order in the tree
+		nodeSnapshot.HavePodsWithAffinityNodeInfoList = make([]*schedulernodeinfo.NodeInfo, 0, cache.nodeTree.numNodes)
```

Comment:
I think this should be switched with Line 280. `NodeInfoList` should only be recreated when updateAll is true.

Reply:
done.